MINOR: Cleanup Server Module

sjhajharia opened the pull request:
As the PR title suggests, this PR is an attempt to perform some cleanups
in the server module. The changes are mostly around the use of Record
type for some classes, changes to use enhanced switch, etc.

Reviewers: Ken Huang <s7133700@gmail.com>, Chia-Ping Tsai
 <chia7712@gmail.com>

sjhajharia committed:
Cleanup Server Module

sjhajharia commented:
Hey @chia7712 
A minor PR for review.

m1a2st reviewed (approved):
Thanks, LGTM

sjhajharia committed:
Empty Commit

chia7712 commented on server/src/main/java/org/apache/kafka/server/share/session/ShareSessionKey.java:
```diff
@@ -21,23 +21,12 @@
-    }
-
     @Override
     public int hashCode() {
```
If the implementation is equal to generated code, please remove them to simplify the code base :)

sjhajharia replied:
The `hashCode()` method is needed if we have an `equals()` method defined. And given that we have a slightly different equals() method than the default one, i think we need to preserve both.

chia7712 replied:
> And given that we have a slightly different equals() method than the default one

pardon me, what is the difference between these two implementations?

sjhajharia replied:
Sorry, I was reading the code a bit wrong. I see the point you are trying to make. I have updated the PR accordingly.

sjhajharia committed:
Merge remote-tracking branch 'origin/trunk' into cleanup-server-module

sjhajharia committed:
add cleanups

sjhajharia committed:
cleanup sharesessionkey

sjhajharia commented:
Gentle reminder @chia7712 
TIA!

sjhajharia committed:
Merge remote-tracking branch 'origin/trunk' into cleanup-server-module

sjhajharia commented:
Gentle reminder @chia7712

sjhajharia committed:
more cleanup

chia7712 commented on server/src/main/java/org/apache/kafka/server/config/ClientQuotaManagerConfig.java:
```diff
@@ -15,6 +15,7 @@
  */
 package org.apache.kafka.server.config;
+
 public class ClientQuotaManagerConfig {
```
Could you please convert it to `record` class?

chia7712 commented on server/src/main/java/org/apache/kafka/server/quota/ClientSensors.java:
```diff
@@ -43,24 +39,12 @@ public ClientSensors(Map<String, String> metricTags,
-                ", quotaSensor=" + quotaSensor +
-                ", throttleTimeSensor=" + throttleTimeSensor +
-                '}';
+            "metricTags=" + metricTags +
```
Do we really need those custom `toString` methods? it seems to me that the generated `toString` is already useful enough

chia7712 commented on server/src/main/java/org/apache/kafka/server/purgatory/DelayedDeleteRecords.java:
```diff
@@ -57,7 +56,7 @@ public DelayedDeleteRecords(long delayMs,
         super(delayMs);
         this.onAcksPending = onAcksPending;
-        this.deleteRecordsStatus = Collections.unmodifiableMap(deleteRecordsStatus);
+        this.deleteRecordsStatus = Map.copyOf(deleteRecordsStatus);
```
Is the deep copy necessary for now? or we could revisit this code after rewriting the `ReplicaManager` by java

sjhajharia replied:
I agree with the later. Lets keep it as Map.copyOf and revisit the same once the Java conversion is complete.

chia7712 commented on server/src/main/java/org/apache/kafka/network/SocketServerConfigs.java:
```diff
@@ -175,15 +173,12 @@ public class SocketServerConfigs {
-        DEFAULT_NAME_TO_SECURITY_PROTO = Collections.unmodifiableMap(nameToSecurityProtocol);
-    }
+    public static final Map<ListenerName, SecurityProtocol> DEFAULT_NAME_TO_SECURITY_PROTO =
+        Map.copyOf(Arrays.stream(SecurityProtocol.values())
```
we could use `toUnmodifiableMap` to avoid redundant deep copy

chia7712 commented on server/src/test/java/org/apache/kafka/network/SocketServerConfigsTest.java:
```diff
@@ -36,7 +36,7 @@ public void testDefaultNameToSecurityProto() {
                 new ListenerName("SASL_SSL"), SecurityProtocol.SASL_SSL
         );
-        assertEquals(expected, SocketServerConfigs.DEFAULT_NAME_TO_SECURITY_PROTO);
+        assertEquals(SocketServerConfigs.DEFAULT_NAME_TO_SECURITY_PROTO, expected);
```
the origin order is already correct, shouldn't it?

chia7712 replied:
oh, that is poor naming. Perhaps, we could simplify the code as follows.
```java
        assertEquals(Map.of(
                new ListenerName("PLAINTEXT"), SecurityProtocol.PLAINTEXT,
                new ListenerName("SSL"), SecurityProtocol.SSL,
                new ListenerName("SASL_PLAINTEXT"), SecurityProtocol.SASL_PLAINTEXT,
                new ListenerName("SASL_SSL"), SecurityProtocol.SASL_SSL
        ), SocketServerConfigs.DEFAULT_NAME_TO_SECURITY_PROTO);
```

chia7712 reviewed (commented):
@sjhajharia thanks for this patch. a couple of comments are left.

sjhajharia committed:
Address comments

chia7712 commented on server/src/main/java/org/apache/kafka/server/Assignment.java:
```diff
@@ -99,13 +57,11 @@ boolean valid(int nodeId, MetadataImage image) {
 
     @Override
     public String toString() {
```
Perhaps we could delegate this task to the Java compiler?

sjhajharia replied:
Updated. Thanks!

chia7712 reviewed (commented):
@sjhajharia thanks for updates.

sjhajharia committed:
celanup assignments

sjhajharia commented:
Gentle ping @chia7712

chia7712 commented on server/src/main/java/org/apache/kafka/server/share/session/ShareSessionKey.java:
```diff
@@ -21,44 +21,16 @@
-        }
-    }
-
     public String toString() {
```
this can be removed

chia7712 commented on server/src/main/java/org/apache/kafka/server/share/session/ShareSessionKey.java:
```diff
@@ -21,44 +21,16 @@
-    private final Uuid memberId;
-
+public record ShareSessionKey(String groupId, Uuid memberId) {
     public ShareSessionKey(String groupId, Uuid memberId) {
```
```java
    public ShareSessionKey {
        Objects.requireNonNull(groupId);
        Objects.requireNonNull(memberId);
    }
```
Also, please add unit test for it

chia7712 commented on server/src/main/java/org/apache/kafka/server/quota/ClientSensors.java:
```diff
@@ -25,14 +25,10 @@
-    private final Sensor quotaSensor;
-    private final Sensor throttleTimeSensor;
-
+public record ClientSensors(Map<String, String> metricTags, Sensor quotaSensor, Sensor throttleTimeSensor) {
```
```java
/**
 * Represents the sensors aggregated per client
 * @param metricTags         quota metric tags for the client
 * @param quotaSensor        sensor that tracks the quota
 * @param throttleTimeSensor sensor that tracks the throttle time
 */
public record ClientSensors(Map<String, String> metricTags, Sensor quotaSensor, Sensor throttleTimeSensor) {
    public ClientSensors {
        metricTags = new LinkedHashMap<>(metricTags);
        Objects.requireNonNull(quotaSensor);
        Objects.requireNonNull(throttleTimeSensor);
    }
}
```
This also needs unit test. Especially, the input order must be kept

chia7712 reviewed (commented):
@sjhajharia thanks for you updates!

sjhajharia committed:
Address comments

sjhajharia commented:
Thanks @chia7712 for the review. 
I have addressed the comments and added two unit test files -> `ClientSensorsTest.java` and `ShareSessionKeyTest.java` for the same.
Pls review when possible. TIA!

sjhajharia committed:
fix spotless errors

sjhajharia committed:
Merge remote-tracking branch 'origin/trunk' into cleanup-server-module

sjhajharia commented:
Hey @chia7712 
gentle reminder
